### PR TITLE
[SPARK-7721][PYTHON][TESTS] Adds PySpark coverage generation script

### PR DESCRIPTION
**Pull request by HyukjinKwon**
## What changes were proposed in this pull request?

Note that this PR was made based on the top of https://github.com/apache/spark/pull/20151. So, it almost leaves the main codes intact.

This PR proposes to add a script for the preparation of automatic PySpark coverage generation. Now, it's difficult to check the actual coverage in case of PySpark. With this script, it allows to run tests by the way we did via `run-tests` script before. The usage is exactly the same with `run-tests` script as this basically wraps it.

This script and PR alone should also be useful. I was asked about how to run this before, and seems some reviewers (including me) need this. It would be also useful to run it manually.

It usually requires a small diff in normal Python projects but PySpark cases are a bit different because apparently we are unable to track the coverage after it's forked. So, here, I made a custom worker that forces the coverage, based on the top of https://github.com/apache/spark/pull/20151.

I made a simple demo. Please take a look - https://spark-test.github.io/pyspark-coverage-site. 

To show up the structure, this PR adds the files as below:

```
python
├── .coveragerc  # Runtime configuration when we run the script.
├── run-tests-with-coverage  # The script that has coverage support and wraps run-tests script.
└── test_coverage  # Directories that have files required when running coverage.
    ├── conf
    │   └── spark-defaults.conf  # Having the configuration 'spark.python.daemon.module'.
    ├── coverage_daemon.py  # A daemon having custom fix and wrapping our daemon.py
    └── sitecustomize.py  # Initiate coverage with COVERAGE_PROCESS_START
```

Note that this PR has a minor nit:

[This scope](https://github.com/apache/spark/blob/04e44b37cc04f62fbf9e08c7076349e0a4d12ea8/python/pyspark/daemon.py#L148-L169) in `daemon.py` is not in the coverage results as basically I am producing the coverage results in `worker.py` separately and then merging it. I believe it's not a big deal.

In a followup, I might have a site that has a single up-to-date PySpark coverage from the master branch as the fallback / default, or have a site that has multiple PySpark coverages and the site link will be left to each pull request.

## How was this patch tested?

Manually tested. Usage is the same with the existing Python test script - `./python/run-tests`. For example,

```
sh run-tests-with-coverage --python-executables=python3 --modules=pyspark-sql
```

Running this will generate HTMLs under `./python/test_coverage/htmlcov`.
  
Console output example:

```
sh run-tests-with-coverage --python-executables=python3,python --modules=pyspark-core
Running PySpark tests. Output is in /.../spark/python/unit-tests.log
Will test against the following Python executables: ['python3', 'python']
Will test the following Python modules: ['pyspark-core']
Starting test(python): pyspark.tests
Starting test(python3): pyspark.tests
...
Tests passed in 231 seconds
Combining collected coverage data under /.../spark/python/test_coverage/coverage_data
Reporting the coverage data at /...spark/python/test_coverage/coverage_data/coverage
Name                         Stmts   Miss Branch BrPart  Cover
--------------------------------------------------------------
pyspark/__init__.py             41      0      8      2    96%
...
pyspark/profiler.py             74     11     22      5    83%
pyspark/rdd.py                 871     40    303     32    93%
pyspark/rddsampler.py           68     10     32      2    82%
...
--------------------------------------------------------------
TOTAL                         8521   3077   2748    191    59%
Generating HTML files for PySpark coverage under /.../spark/python/test_coverage/htmlcov
```

### CODE REVIEW

File: python/run-tests.py [hunk 1]
```diff
@@ -175,6 +175,9 @@ def main():
 
     task_queue = Queue.PriorityQueue()
     for python_exec in python_execs:
+        if "COVERAGE_PROCESS_START" in os.environ:
+            # Make sure if coverage is installed.
+            run_cmd([python_exec, "-c", "import coverage"])
```
Comment (HyukjinKwon) [inline python/run-tests.py]: Manually tested:

```
./run-tests-with-coverage --python-executables=python3
Running PySpark tests. Output is in /.../spark/python/unit-tests.log
Will test against the following Python executables: ['python3']
Will test the following Python modules: ['pyspark-core', 'pyspark-ml', 'pyspark-mllib', 'pyspark-sql', 'pyspark-streaming']
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ModuleNotFoundError: No module named 'foo'
[error] running python3 -c import foo ; received return code 1
```

**Comment by HyukjinKwon** [comment]
cc @rxin, @felixcheung, @ueshin, @holdenk and @JoshRosen. Here, I made it based on #20151. Could you take a look for this one too? I can make this PR separate from #20151 by manually adding the custom fix into `worker.py` if requested.

**Comment by HyukjinKwon** [comment]
cc @icexelloss too. I think we had a small talk about this before.

**Comment by icexelloss** [comment]
This is awesome. Thanks @HyukjinKwon!

**Comment by HyukjinKwon** [comment]
Sorry for noice. I made some clean up and just rebased.

**Comment by icexelloss** [comment]
I usually run tests like this:
```
SPARK_TESTING=1 bin/pyspark pyspark.sql.tests ...
```
Does it make sense to expose similar usage for enabling coverage? Sth like
```
SPARK_TESTING_WITH_COVERAGE=1 bin/pyspark pyspark.sql.tests ...
```

**Review by felixcheung — COMMENTED** [review]
looks reasonable to me.
how are we going to track changes to coverage number?

**Comment by HyukjinKwon** [comment]
If anyone is worried of elapsed time, I think seems generally fine:

Before:

```
sh run-tests --python-executables=python3 --modules=pyspark-sql
```

```
...
Finished test(python3): pyspark.sql.readwriter (34s)
Finished test(python3): pyspark.sql.tests (241s)
Tests passed in 241 seconds
```

After:

```
sh run-tests-with-coverage  --python-executables=python3 --modules=pyspark-sql
```

```
...
Finished test(python3): pyspark.sql.conf (288s)
Finished test(python3): pyspark.sql.tests (288s)
Tests passed in 288 seconds
```

Given full test usually take `Tests passed in 1103 seconds`, I think it will only increase 2~3 more minutes even if we run this with all Python executables.

**Comment by HyukjinKwon** [comment]
@icexelloss, for https://github.com/apache/spark/pull/20204#issuecomment-356431026, yes, that's the way I usually use too. My worry is though I wonder if this is a proper official way to do it because I have been thinking this way is rather meant to be internal. 

For this reason, I left a comment in this script for now:

```diff
+# If you'd like to run a specific unittest class, you could do such as
+# SPARK_TESTING=1 ../bin/pyspark pyspark.sql.tests VectorizedUDFTests
+./run-tests $@
```

If I remember this correctly, I think I also had a short talk about it with @nchammas and @BryanCutler before (I think it was about more detailed control tho).

I think we should better fix `run_tests` script to accept the unittest class as an option. I took a look this before and I guess it won't be too difficult to introduce another option. Once we fix it, it will also be available in this script too because the script here wraps `run_tests`. 

Will probably try to take a look again and open another PR separately (maybe within the following week?)

**Comment by HyukjinKwon** [comment]
> looks reasonable to me.
> how are we going to track changes to coverage number?

Yea, that's a good point. I would like to do it too because other tools allow it. I think this is another place where I should investigate more when I actually give a shot to integrate it with Jenkins. Will cc you in the JIRA, share my investigation, and try to do this in the followup I mentioned in the PR description.

File: python/test_coverage/coverage_daemon.py [hunk 2]
```diff
@@ -0,0 +1,45 @@
...
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import os
+import imp
+
+
+# This is a hack to always refer the main code rather than built zip.
```
Comment (HyukjinKwon) [inline python/test_coverage/coverage_daemon.py]: So, we will always track the coverage in the main code under `python/pyspark`, not the built zip Python library.

File: python/test_coverage/sitecustomize.py [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+
+import coverage
+coverage.process_startup()
```
Comment (BryanCutler) [inline python/test_coverage/sitecustomize.py]: would you mind explaining how this file is called and used?  I can't quite figure it out..

**Review by BryanCutler — COMMENTED** [review]
This is great, thanks for working on it!

File: python/test_coverage/sitecustomize.py (hunk 3 above)
Comment (HyukjinKwon) [reply python/test_coverage/sitecustomize.py]: Yup, will add some comments.

**Comment by icexelloss** [comment]
> @icexelloss, for #20204 (comment), yes, that's the way I usually use too. My worry is though I wonder if this is a proper official way to do it because I have been thinking this way is rather meant to be internal.

Sounds good to me. Thanks!

**Comment by HyukjinKwon** [comment]
retest this please

File: python/run-tests-with-coverage [hunk 4]
```diff
@@ -0,0 +1,69 @@
...
+
+# We use 'spark.python.daemon.module' configuration to insert the coverage supported workers.
+export SPARK_CONF_DIR="$COVERAGE_DIR/conf"
+
+# This environment variable enables the coverage.
+export COVERAGE_PROCESS_START="$FWDIR/.coveragerc"
+
+# If you'd like to run a specific unittest class, you could do such as
+# SPARK_TESTING=1 ../bin/pyspark pyspark.sql.tests VectorizedUDFTests
+./run-tests $@
```
[the remark this answers is not in the document]
Comment (HyukjinKwon) [reply python/run-tests-with-coverage]: Yeap, will update.

File: python/run-tests-with-coverage [hunk 5]
```diff
@@ -0,0 +1,69 @@
...
+#
+
+set -o pipefail
+set -e
+
+# This variable indicates which coverage executable to run to combine coverages
+# and generate HTMLs, for example, 'coverage3' in Python 3.
+COV_EXEC="${COV_EXEC:-coverage}"
+FWDIR="$(cd "`dirname $0`"; pwd)"
+pushd "$FWDIR" > /dev/null
```
Comment (ueshin) [inline python/run-tests-with-coverage]: Do we need to use `pushd` and its corresponding `popd` at the end of this file? I guess we can simply use `cd` here.

**Review by ueshin — COMMENTED** [review]
LGTM except for one comment.

File: python/run-tests-with-coverage (hunk 5 above)
Comment (felixcheung) [reply python/run-tests-with-coverage]: my 2 c: I think it's ok, I'd prefer it; might be useful in the future when more cd are added

**Review by felixcheung — APPROVED** [review]
LGTM
perhaps great to follow up with some steps on how to run it, what to look for, how to diff changes :) 
but this is a very dev centric thing and you have a lot of comments in-line, so I'd say we are good here.

File: python/run-tests-with-coverage (hunk 5 above)
Comment (ueshin) [reply python/run-tests-with-coverage]: I see, no problem at all. I just wanted to confirm. Thanks!

**Comment by HyukjinKwon** [comment]
retest this please

File: python/run-tests-with-coverage [hunk 6]
```diff
@@ -0,0 +1,69 @@
...
+
+# We use 'spark.python.daemon.module' configuration to insert the coverage supported workers.
+export SPARK_CONF_DIR="$COVERAGE_DIR/conf"
+
+# This environment variable enables the coverage.
+export COVERAGE_PROCESS_START="$FWDIR/.coveragerc"
+
+# If you'd like to run a specific unittest class, you could do such as
+# SPARK_TESTING=1 ../bin/pyspark pyspark.sql.tests VectorizedUDFTests
+./run-tests "$@"
```
Comment (HyukjinKwon) [inline python/run-tests-with-coverage]: Another tip is, if we use `../bin/pyspark` here, do some simple tests and then exit, it looks still producing the coverage correctly.

**Comment by HyukjinKwon** [comment]
Will merge this one if there's no more comments in few days.

**Comment by HyukjinKwon** [comment]
Merged to master.

**Pull request closed without merging**